Fixes a race condition in IabHelper.

kplatfoot opened the pull request:
The IabHelper class has a dispose() method that does various cleanup tasks
(unbinds the ServiceConnection, nulls out instance fields, etc). This method
is intended to be called from the host activity's onDestroy method. IabHelper
also includes *Async methods that can perform billing operations on a back-
ground thread. However, this combination produces a race condition that can
cause runtime crashes. If dispose() is called while an async operation is
in progress, the async operation may crash, e.g. because mDisposed is false
(IllegalStateException) or mContext is null (NullPointerException).

The fix is to have dispose() check whether an async operation is in progress,
and if so, defer the cleanup until after it finishes.

kplatfoot commented:
I signed the CLA.

kplatfoot commented:
Friendly ping for a review of my fix commit (https://github.com/kplatfoot/android-play-billing/commit/d6e39b1ba339c5f0225adc0d301237066a38e2cd). Thanks!

PaulRashidi commented:
Hi, sorry for not responding sooner. Just curious why you preferred to join the dispose thread with the async one? Perhaps giving the caller a choice of disposeNow() or disposeDeferred() would be better?

kplatfoot commented:
No problem, Paul.

My patch doesn't actually join the dispose thread to the async thread (i.e. the dispose thread is not blocked until the async thread finishes). It's just setting a flag (mDisposeAfterAsync) to ensure that dispose() is eventually called when it's safe to do so (after the async op is done).

If IabHelper provided a disposeNow() method, calling it would be just as dangerous during async operations as dispose() is currently. Dispose() nulls out fields (mContext, mService) that other methods (e.g. queryInventory) assume are set, resulting in random NPEs whenever IabHelper is disposed with a pending async op. The class Javadoc recommends disposing IabHelper when its host Activity is destroyed, so if, for example, you call queryInventoryAsync and the user rotates the screen immediately afterward (triggering onDestroy), there's a good chance the app will crash.

If there were separate disposeNow() and disposeDeferred() methods, the user would need to track whether there was a pending async operation and be sure to call disposeDeferred() if so. Since IabHelper already knows this (via mAsyncInProgress), it seems more straightforward to make the deferred dispose an implementation detail rather than part of the public API.

PaulRashidi commented:
Sorry about the join comment, was a little late. The dispose pattern seems to typically account for freeing resources when garbage collection is present so it might be misleading to silently defer the dispose work as an implementation detail. Seems like this code should interrupt the in progress work or surface the deferred nuance through changing the method name and javadoc to something like disposeWhenComplete().

kplatfoot commented:
Thanks for the clarification, Paul.

I believe having dispose() interrupt the in-progress async operation would be messy. This would require every method that can be invoked from the background thread (queryInventory, querySkuDetails, queryPurchases, etc) to guard every usage of mService and mContext with null checks, just in case dispose() was called from another thread.

Here's a proposal: What if dispose() simply checked mAsyncInProgress and threw an IllegalStateException if there was a pending async op? This would match the semantics of the existing async methods (launchPurchaseFlow, queryInventoryAsync, etc), which cannot be called until the previous async op is finished. It would still be an improvement over the old code, because you would get a clear exception immediately (e.g. "Can't dispose because an async operation (x) is in progress.") rather than a random, mysterious NPE from whatever method happened to be running on the background thread.

For developer convenience, I think it's worth keeping the new dispose method but with a different name, like you suggested. The Javadoc would recommend using the new method if it's possible that an async operation is in progress.

PaulRashidi commented:
I agree the interrupt is a bit more work.

Your proposals sounds good. One possible tweak: given the intended use case of adding dispose to Activity.destroy() we might want to make it a checked exception.

kplatfoot commented:
Thanks, Paul. Yeah, I can make dispose() throw a checked exception to ensure that developers are aware of the possibility and handle it. Do you think it would make sense to have flagStartAsync throw the same exception (instead of the IllegalStateException it does now), since the semantics are basically the same (invalid to call while another async op is in progress)?

PaulRashidi commented:
I think that's also a safe move. The dev can choose to ignore, but it would be an explicit and documented decision.

kplatfoot commented:
Ok, I've updated my patch (https://github.com/kplatfoot/android-play-billing/commit/e17a11c0987b0969f1c66d6684acc793c5c17ed6) to have dispose() and all async methods throw a checked exception if there's already an async op in progress. I've also added a new method, disposeWhenFinished(), which queues the dispose to happen after any in-progress op is complete. Using a checked exception made the commit somewhat larger, since I needed to update many method signatures in IabHelper and calls from MainActivity. Let me know what you think. Thanks!

jkane001 commented:
@kplatfoot Is there someone you can assign this to? It seems to have been ignored so far...

kplatfoot commented:
@jkane001 I'm hoping it will be. I believe I've addressed all the suggestions Paul has made.

kplatfoot commented:
Paul, can I get another look at this CL?

BTW, my other pull request (https://github.com/googlesamples/android-play-billing/pull/11, from March 7) is also still awaiting review.

Thanks!
